### ISSUE TITLE
[JENKINS-55485] Declarative pipeline, lock() in stage options is executed before when clause

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/lockable-resources-plugin#916]
<p>In a a declarative pipeline i have a stage grouping several other stages and need to hold a lock for the duration of the group.  I see I can add lock() to the options of the group and it works fine in that scenario.  However the group also has a When clause so I can turn that group off with a job parameter.  </p>

<p>The problem is that when the "when" evaluates false so the stage will be skipped, it wait for the lock anyway.  If the stage will be skipped for any reason, we should not block on the lock.  The "when" defines a "beforeAgent" attribute for a similar need, so that might be an approach, but I'm not seeing a scenario where you would want to do it in the current order.</p>

<p> </p>

<p>My pipeline looks like:</p>

<p>—</p>

<p>pipeline {</p>

<p>agent none<br/>
 stages {</p>

<p>      stage ('Build')</p>

{      }

<p>      stage('Deploy-and-Test') { // Group Stage Deploy and test the Build - lock the deployment VM so we don't step on other runs<br/>
           when {<br/>
                expression {<br/>
                     "${params.Deploy}" == "true"<br/>
                }<br/>
               beforeAgent true<br/>
           }<br/>
           options {<br/>
                lock(quantity: 1, resource: "${params.DeployAgent}", variable: 'myDeployAgent')<br/>
           }<br/>
           agent none</p>

<p>          stages { ...}</p>

<p>     }}}</p>

<p> </p>

<p>When I run it withthe param to disable the stage I see:</p>

<p>—<br/>
 <span class="error">&#91;Pipeline&#93;</span> stage<span class="error">&#91;Pipeline&#93;</span> { (Deploy-and-Test)<span class="error">&#91;Pipeline&#93;</span> lockTrying to acquire lock on <span class="error">&#91;DeploymentVM_Win&#93;</span><br/>
 Lock acquired on <span class="error">&#91;DeploymentVM_Win&#93;</span><span class="error">&#91;Pipeline&#93;</span> {Stage "Deploy-and-Test" skipped due to when conditional<br/>
 —</p>

<p>Since the stage is being skipped there is no point in holding the lock, but it the lock is already allocated, this will block anyway.  The lock() should be processed <em>after</em> the decision whether to execute the stage.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wgc123">wgc123</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-55485" target="_blank">Declarative pipeline, lock() in stage options is executed before when clause</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: lockable-resources-plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2019-10-05T01:12:05+00:00
<li><b>votes</b>: 3
<li><b>watchers</b>: 7
<li><b>imported</b>: 20251217-081411
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>In a a declarative pipeline i have a stage grouping several other stages and need to hold a lock for the duration of the group.  I see I can add lock() to the options of the group and it works fine in that scenario.  However the group also has a When clause so I can turn that group off with a job parameter.  </p>

<p>The problem is that when the "when" evaluates false so the stage will be skipped, it wait for the lock anyway.  If the stage will be skipped for any reason, we should not block on the lock.  The "when" defines a "beforeAgent" attribute for a similar need, so that might be an approach, but I'm not seeing a scenario where you would want to do it in the current order.</p>

<p> </p>

<p>My pipeline looks like:</p>

<p>—</p>

<p>pipeline {</p>

<p>agent none
 stages {</p>

<p>      stage ('Build')</p>

{      }

<p>      stage('Deploy-and-Test') { // Group Stage Deploy and test the Build - lock the deployment VM so we don't step on other runs
           when {
                expression {
                     "${params.Deploy}" == "true"
                }
               beforeAgent true
           }
           options {
                lock(quantity: 1, resource: "${params.DeployAgent}", variable: 'myDeployAgent')
           }
           agent none</p>

<p>          stages { ...}</p>

<p>     }}}</p>

<p> </p>

<p>When I run it withthe param to disable the stage I see:</p>

<p>—
 <span class="error">&#91;Pipeline&#93;</span> stage<span class="error">&#91;Pipeline&#93;</span> { (Deploy-and-Test)<span class="error">&#91;Pipeline&#93;</span> lockTrying to acquire lock on <span class="error">&#91;DeploymentVM_Win&#93;</span>
 Lock acquired on <span class="error">&#91;DeploymentVM_Win&#93;</span><span class="error">&#91;Pipeline&#93;</span> {Stage "Deploy-and-Test" skipped due to when conditional
 —</p>

<p>Since the stage is being skipped there is no point in holding the lock, but it the lock is already allocated, this will block anyway.  The lock() should be processed <em>after</em> the decision whether to execute the stage.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>wgc123:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-55485?focusedId=358157&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-358157">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I verified the workaround of adding an additional level of stage - I have a group stage with the when clause containing a group stage with the lock, containing the actual stages.  With this structure I'm able to use the when clause to disable deploy and test stages, and not get stuck waiting for a lock behind a long-running test.</p>

<p>The workaround is conceptually easy, but adds a lot of boilerplate and makes the Jenkinsfile that much harder to read.  I'm now up to 10 levels of tabs in my Jenkinsfile and have a check-in with hundreds of lines of changes that is just indenting.  It's getting pretty ugly.</p>

<p>This is exactly the mess the "beforeAgent" directive to the "when" clause prevents - we need a similar solution here.</p>
</pre>
</details>

</ul>
</details>

<p>I verified the workaround of adding an additional level of stage - I have a group stage with the when clause containing a group stage with the lock, containing the actual stages.  With this structure I'm able to use the when clause to disable deploy and test stages, and not get stuck waiting for a lock behind a long-running test.</p>

<p>The workaround is conceptually easy, but adds a lot of boilerplate and makes the Jenkinsfile that much harder to read.  I'm now up to 10 levels of tabs in my Jenkinsfile and have a check-in with hundreds of lines of changes that is just indenting.  It's getting pretty ugly.</p>

<p>This is exactly the mess the "beforeAgent" directive to the "when" clause prevents - we need a similar solution here.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/amrichko.png" title="amrichko's avatar" /> amrichko:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-55485?focusedId=361050&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-361050">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Also faced this issue and want to add that it is really painful in case of concurrent builds that use the same lock resource.</p>

<p>For example:</p>

<p>Build 1::  Stage 1 ----------&gt; Stage 2 (get lock A) <span class="error">&#91;(when branch is master) == true &#93;</span> -----------&gt; Stage 3</p>

<p>Build 2::  Stage 1 ----------&gt; Stage 2 (get lock A) <span class="error">&#91;(when branch is master) == false &#93;</span> -----------&gt; Stage 3</p>

<p> </p>

<p><b>Expected behaviour:</b></p>

<p>Build 2 skips Stage 2 because it executed not from master branch. Lock not even tried to be allocated</p>

<p> </p>

<p><b>Current  behaviour:</b></p>

<p>Build 2 is trying to allocate lock A before checking when conditional, so it will skip Stage 2 only after Stage 2 will be finished in Build 1</p>
</pre>
</details>

</ul>
</details>

<p>Also faced this issue and want to add that it is really painful in case of concurrent builds that use the same lock resource.</p>

<p>For example:</p>

<p>Build 1::  Stage 1 ----------> Stage 2 (get lock A) <span class="error">&#91;(when branch is master) == true &#93;</span> -----------> Stage 3</p>

<p>Build 2::  Stage 1 ----------> Stage 2 (get lock A) <span class="error">&#91;(when branch is master) == false &#93;</span> -----------> Stage 3</p>

<p> </p>

<p><b>Expected behaviour:</b></p>

<p>Build 2 skips Stage 2 because it executed not from master branch. Lock not even tried to be allocated</p>

<p> </p>

<p><b>Current  behaviour:</b></p>

<p>Build 2 is trying to allocate lock A before checking when conditional, so it will skip Stage 2 only after Stage 2 will be finished in Build 1</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>haw777:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-55485?focusedId=364358&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364358">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is this a duplicate of 51865?</p>
</pre>
</details>

</ul>
</details>

<p>Is this a duplicate of 51865?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>wgc123:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-55485?focusedId=372721&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-372721">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, this duplicates:  <a href="https://issues.jenkins.io/browse/JENKINS-51865" title="Stage locks are created for skipped stages in declarative pipeline" class="issue-link" data-issue-key="JENKINS-51865"><del>JENKINS-51865</del></a></p>
</pre>
</details>

</ul>
</details>

<p>Yes, this duplicates:  <a href="https://issue-redirect.jenkins.io/issue/51865" title="Stage locks are created for skipped stages in declarative pipeline" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/51865">JENKINS-51865</a>"><del>JENKINS-51865</del></a></p>
